### PR TITLE
[RSDK-8602] Release stream image after for loop termination (Gostream) 

### PR DESCRIPTION
**Pull request by hexbabe**
[RSDK-8602](https://viam.atlassian.net/browse/RSDK-8602)

[RSDK-8602]: https://viam.atlassian.net/browse/RSDK-8602?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: gostream/media.go [hunk 1]
```diff
@@ -313,11 +316,14 @@ func (pc *producerConsumer[T, U]) start() {
 					}
 				}, err}
 				pc.currentMu.Unlock()
-				if lastRelease != nil {
-					lastRelease()
-				}
 			}()
 		}
+		if lastRelease != nil {
+			lastRelease()
```
Comment (dgottlieb) [inline gostream/media.go:322]: Why do we believe this hasn't been called?

File: gostream/media.go [hunk 2]
```diff
@@ -284,7 +285,9 @@ func (pc *producerConsumer[T, U]) start() {
 					doReadSpan.End()
 				}()
 
-				var lastRelease func()
+				if lastRelease != nil {
```
Comment (dgottlieb) [inline gostream/media.go:288]: What was wrong about where `lastRelease` was before?

Comment (hexbabe) [reply gostream/media.go:288]: Good point. If I move it back to where it was. We also don't need the lastRelease() after the loop

File: gostream/media.go [hunk 3]
```diff
@@ -313,11 +313,14 @@ func (pc *producerConsumer[T, U]) start() {
 					}
 				}, err}
 				pc.currentMu.Unlock()
-				if lastRelease != nil {
```
Comment (hexbabe) [inline gostream/media.go:316]: `lastRelease` "last" is ambiguously similar to "final" and confused me a bit on first glance

File: gostream/media_test.go [hunk 4]
```diff
@@ -13,70 +13,175 @@ import (
...
+// WrappedImage wraps an image.Image and includes a bool flag to track release status.
+type WrappedImage struct {
+	Image    image.Image
+	released bool
+	mu       sync.Mutex
+}
+
+// Implement the image.Image interface for WrappedImage.
+func (wi *WrappedImage) ColorModel() color.Model {
+	return wi.Image.ColorModel()
```
[the remark this answers is not in the document]
Comment (hexbabe) [reply gostream/media_test.go:31]: Added something to address this

**Comment by hexbabe** [comment]
@randhid Dan is on PTO and I would like to merge this in before the RC gets cut. Could you help review? Thankyouu

**Comment by randhid** [comment]
Tagging mr gostream himself

File: gostream/media.go [hunk 5]
```diff
@@ -313,11 +313,14 @@ func (pc *producerConsumer[T, U]) start() {
...
-					lastRelease()
+				if prevRelease != nil {
+					prevRelease()
 				}
 			}()
 		}
+		if pc.current != nil && pc.current.Release != nil {
+			pc.current.Release()
+		}
```
Comment (randhid) [inline gostream/media.go:323]: Add a comment above this explaining the need for releasing the current producer-consumer if it exists.

**Review by randhid — APPROVED** [review]
Looks plausible to me, so the defer function after the managed go was insufficient to release the image from the producer/consumer loop?

**Comment by hexbabe** [comment]
> Looks plausible to me, so the defer function after the managed go was insufficient to release the image from the producer/consumer loop?

If my understanding is correct, `defer pc.activeBackgroundWorkers.Done(); pc.cancel()` would only decrement the managed go stuff and invoke a context cancel. Nothing happens with handling `pc.current` so we need that extra cleanup

**Comment by hexbabe** [comment]
> > Looks plausible to me, so the defer function after the managed go was insufficient to release the image from the producer/consumer loop?
> 
> If my understanding is correct, `defer pc.activeBackgroundWorkers.Done(); pc.cancel()` would only decrement the managed go stuff and invoke a context cancel. Nothing happens with handling `pc.current` so we need that extra cleanup

It may be possible to move the new cleanup current logic into a defer anonymous func for clarity? 90% sure that would work?

**Comment by randhid** [comment]
> > > Looks plausible to me, so the defer function after the managed go was insufficient to release the image from the producer/consumer loop?
> > 
> > 
> > If my understanding is correct, `defer pc.activeBackgroundWorkers.Done(); pc.cancel()` would only decrement the managed go stuff and invoke a context cancel. Nothing happens with handling `pc.current` so we need that extra cleanup
> 
> It may be possible to move the new cleanup current logic into a defer anonymous func for clarity? 90% sure that would work?

It's okay to do this in a follow up pr, plus the atomic bools dan mentioned, if you want to get this in asap.

File: gostream/media.go [hunk 6]
```diff
@@ -313,11 +313,15 @@ func (pc *producerConsumer[T, U]) start() {
...
 				}, err}
 				pc.currentMu.Unlock()
-				if lastRelease != nil {
-					lastRelease()
+				if prevRelease != nil {
+					prevRelease()
 				}
 			}()
 		}
+		// After loop breaks, we likely still have an unreleased current media.
```
Comment (edaniels) [inline gostream/media.go:321]: nit: likely -> may

**Review by edaniels — APPROVED** [review]
This seems fine!

**Pull request merged**